### ISSUE TITLE
Add except: ["first-nested-and-blockless"] option to at-rule-empty-line-before

### ISSUE THREAD

**Issue by AlecRust** [stylelint/stylelint#2480]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

If this rule checked for `@media` or perhaps an end brace:

```css
@media screen and (min-width: 480px) {
```

It would allow syntax like this not to need to be ignored with `/* stylelint-disable-line */`.

> Which rule, if any, is this issue related to?

at-rule-empty-line-before

> What CSS is needed to reproduce this issue?

```css
.foo {
  @inherit: .bar;
  height: 40px;
}
```

> What stylelint configuration is needed to reproduce this issue?
```json
{
  "rules": {
    "declaration-no-important": "always"
  }
}

```

> Which version of stylelint are you using?

7.10.1

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint styles/**/*.css`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, it's related to [postcss-inherit](https://github.com/GarthDB/postcss-inherit)

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warnings were flagged:

![screenshot](https://i.imgur.com/TcvpdQk.png)

**Comment by hudochenkov** [comment]
Please use issue template for describing your problem.

**Comment by AlecRust** [comment]
@hudochenkov apologies, updated.

**Comment by hudochenkov** [comment]
@AlecRust thank you, but you didn't provide your config :)

**Comment by AlecRust** [comment]
@hudochenkov I just included the relevant rule of a config. My actual config is as follows, which sets this rule via [stylelint-config-suitcss](https://github.com/suitcss/stylelint-config-suitcss/blob/master/index.js#L7-L9).

```json
{
  "extends": "stylelint-config-suitcss",
  "plugins": [
    "stylelint-selector-bem-pattern"
  ],
  "rules": {
    "plugin/selector-bem-pattern": {
      "preset": "suit"
    },
    "block-no-empty": null,
    "declaration-no-important": true
  }
}
```

**Comment by jeddy3** [comment]
@AlecRust Thanks for updating your issue to use the template. 

The rule is behaving as expected. `stylelint-config-suitcss` configures the `at-rule-empty-line-before` rule as:

```json
{
 "at-rule-empty-line-before": [ "always", {
    "except": ["blockless-after-blockless"],
  } ],
```

In the follow code, the `@inherit` at-rule does not have an empty line before it, nor is it a blockless at-rules preceding another blockless at-rule. As such, it correctly triggers a warning.

```css
.foo {
  @inherit: .bar;
  height: 40px;
}
```

The `stylelint-config-suitcss` config isn't designed to work with non-standard language extensions like `@inherit`. As such, you'll need to extend the config and override the `at-rule-empty-line-before` configuration with something specific to your needs. You'll want look into the [various `"except"`](https://stylelint.io/user-guide/rules/at-rule-empty-line-before/#except-after-same-name-inside-block-blockless-after-same-name-blockless-blockless-after-blockless-first-nested) options. Let us know how you get on and if one of those options allows you to lint your desired code style. If not, we can work out what option we need to add to add to the rule make it happen :)

**Comment by AlecRust** [comment]
@jeddy3 thanks for your response. Yup am aware the rule is working as expected, I guess my query is whether the rule should be more specific i.e. check for `@media`, `@import`, `@extend` etc. rather than anything beginning with `@`.

Unfortunately I don't see an optional secondary option which would throw no error for my example above but keep the enforcement for `@media` CSS e.g.

```css
.foo {
  display: block;
  
  @media (--bp-mobile) {
    color: red;
  }
}
```

**Comment by jeddy3** [comment]
> thanks for your response.

No worries.

> I guess my query is whether the rule should be more specific i.e. check for `@media`, `@import`, `@extend` etc. rather than anything beginning with `@`.

Yep, that makes sense. In stylelint a rule is applicable to the entire _construct_ by default but exceptions can be specified with the `except*` and `ignore*` options. This is because things like _at-rules_ are, along with _functions_, a syntactic constructs that allow for the extension of the CSS language and, as such, it not viable to have a separate rule for each at-rule type.

> Unfortunately I don't see an optional secondary option which would throw no error for my example above but keep the enforcement for `@media` CSS e.g.

OK, let's work out what's missing. How would you structure the following code snippet when using your code style?:

```css
@charset "UTF8";

@import "x.css";
@import "y.css";

@font-face {
  font-family: X;
  src: "x.woff"
}

.foo {
  @media (--bp-mobile) {
    color: red;
  }
}

.bar {
  @inherit .bar;
  @apply --bar;
  color: blue;
  
  @support (display: flex) {
    @media (--bp-mobile) {
      display: flex;
    }
  }
}
```

Depending on your approach we'll know whether it's better to introduce a new `"except"` option like `"inside-block-and-blockless"` (similar to [here](https://stylelint.io/user-guide/rules/rule-empty-line-before/#inside-block-and-after-rule)) or introduce an `"exceptAtRules": [""]` option that will let you specify which at-rules the primary option should be inverted on.

**Comment by evilebottnawi** [comment]
@jeddy3 I was just about to raise this issue, a new option sgtm to me.

**Comment by evilebottnawi** [comment]
@jeddy3 btw, it is standard code style recommended `sass` (`scss`) comunity

**Comment by AlecRust** [comment]
I'm adhering to `stylelint-config-suitcss` where possible, with the two exceptions you can see in my config above.

So for that example to adhere it would be:

```css
@charset "UTF8";
@import "x.css";
@import "y.css";

@font-face {
  font-family: X;
  src: "x.woff";
}

.foo {

  @media (--bp-mobile) {
    color: red;
  }
}

.bar {
  @inherit .bar;
  @apply --bar;
  color: blue;

  @support (display: flex) {

    @media (--bp-mobile) {
      display: flex;
    }
  }
}
```

In this example `@inherit` (and now `@apply`) are throwing the error. I'm currently working around it with:

```css
.foo {
  @inherit: .bar; /* stylelint-disable-line */
  color: red;
}
```

I was expecting the at-rules with brackets to throw an error when there's no empty line, rather than the property at-rules. I understand now that there can be valid property at-rules. So yes, maybe an option to exclude these property ones?

**Comment by jeddy3** [comment]
OK, I had to think about this. @AlecRust For your code style I think a new `"first-nested-and-blockless"` option is needed, one that is more specific than the `"first-nested"` one.

Options _either_ describe the relationship between the construct and the thing that precedes it, _or_ they described _the thing itself_. This new option is one of the latter.

Then I think the following config would enforce your code style:

```json
{
 "at-rule-empty-line-before": [ "always", {
    "except": [
      "blockless-after-blockless",
      "first-nested-and-blockless"
    ],
  } ]
}
```

@AlecRust I'm going to label this as `help wanted`. Please consider [contributing](https://stylelint.io/developer-guide/rules/#adding-an-option-to-an-existing-rule) this option.

@evilebottnawi Please create a separate issue, as I think your preferred code style differs from @AlecRust's.

**Comment by jeddy3** [comment]
Closing [as stylistic rules are frozen](https://github.com/stylelint/stylelint/issues/5674). The community is welcome to migrate the rule [to a plugin](https://stylelint.io/developer-guide/plugins) and add this option.
